src/Microsoft.DotNet.Watcher.Tools/Internal/MsBuildFileSetFactory.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Watcher.Tools;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.Watcher.Internal
+{
+    public class MsBuildFileSetFactory : IFileSetFactory
+    {
+        private const string ProjectExtensionFileExtension = ".dotnetwatch.targets";
+        private const string WatchTargetsFileName = "DotNetWatchCommon.targets";
+        private readonly ILogger _logger;
+        private readonly string _projectFile;
+        private readonly string _watchTargetsDir;
+
+        public MsBuildFileSetFactory(ILogger logger, string projectFile)
+        {
+            Ensure.NotNull(logger, nameof(logger));
+            Ensure.NotNullOrEmpty(projectFile, nameof(projectFile));
+
+            _logger = logger;
+            _projectFile = projectFile;
+            _watchTargetsDir = FindWatchTargetsDir();
+        }
+
+        public async Task<IFileSet> CreateAsync(CancellationToken cancellationToken)
+        {
+            EnsureInitialized();
+
+            var watchList = Path.GetTempFileName();
+            try
+            {
+                var projectDir = Path.GetDirectoryName(_projectFile);
+
+                while (true)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var msBuildOutput = new List<string>();
+
+#if DEBUG
+                    var stopwatch = new Stopwatch();
+                    stopwatch.Start();
+#endif
+
+                    // TODO adding files doesn't currently work. Need to provide a way to detect new files
```

Comment:
We'll have to rescan the project for files and see if the new file is included in the new set. We had to do the same thing w/ project.json.